src/rust/ide/src/transport/web.rs

```diff
@@ -98,60 +103,198 @@ impl State {
 
 
 // =================
-// === WebSocket ===
+// === JS Events ===
 // =================
 
-/// Wrapper over JS `WebSocket` object and callbacks to its signals.
-#[derive(Debug)]
-pub struct WebSocket {
-    #[allow(missing_docs)]
-    pub logger     : Logger,
-    /// Handle to the JS `WebSocket` object.
-    pub ws         : web_sys::WebSocket,
-    /// Handle to a closure connected to `WebSocket.onmessage`.
-    pub on_message : OptionalFmMutClosure<MessageEvent>,
-    /// Handle to a closure connected to `WebSocket.onclose`.
-    pub on_close   : OptionalFmMutClosure<CloseEvent>,
-    /// Handle to a closure connected to `WebSocket.onopen`.
-    pub on_open    : OptionalFmMutClosure<Event>,
-    /// Handle to a closure connected to `WebSocket.onerror`.
-    pub on_error   : OptionalFmMutClosure<Event>,
+/// Description of events that can be emitted by JS WebSocket.
+pub mod event {
+    use super::*;
+    use ensogl_system_web::event::Event;
+
+    /// Represents WebSocket.open event.
+    #[derive(Clone,Copy,Debug)]
+    pub enum Open{}
+    impl Event for Open {
+        type Interface = web_sys::Event;
+        type Target = web_sys::WebSocket;
+        const NAME:&'static str = "open";
+    }
+
+    /// Represents WebSocket.close event.
+    #[derive(Clone,Copy,Debug)]
+    pub enum Close{}
+    impl Event for Close {
+        type Interface = web_sys::CloseEvent;
+        type Target = web_sys::WebSocket;
+        const NAME:&'static str = "close";
+    }
+
+    /// Represents WebSocket.message event.
+    #[derive(Clone,Copy,Debug)]
+    pub enum Message{}
+    impl Event for Message {
+        type Interface = web_sys::MessageEvent;
+        type Target = web_sys::WebSocket;
+        const NAME:&'static str = "message";
+    }
+
+    /// Represents WebSocket.error event.
+    #[derive(Clone,Copy,Debug)]
+    pub enum Error{}
+    impl Event for Error {
+        type Interface = web_sys::Event;
+        type Target = web_sys::WebSocket;
+        const NAME:&'static str = "error";
+    }
 }
 
-impl WebSocket {
+
+
+// =============
+// === Model ===
+// =============
+
+/// An owning wrapper over JS `WebSocket` object and callbacks to its signals.
+#[derive(Derivative)]
+#[derivative(Debug)]
+#[allow(missing_docs)]
+pub struct Model {
+    // === User-provided callbacks ===
+    pub on_close          : Slot<event::Close>,
+    pub on_message        : Slot<event::Message>,
+    pub on_open           : Slot<event::Open>,
+    pub on_error          : Slot<event::Error>,
+
+
+    // === Internal ===
+    pub logger            : Logger,
+    pub ws                : web_sys::WebSocket,
+    /// Special callback on "close" event. Must be invoked after `on_close`.
+    pub on_close_internal : Slot<event::Close>,
+    /// When enabled, the WS will try to automatically reconnect whenever connection is lost.
+    pub reconnect         : bool,
```

Comment:
The name does not tell what docs tell. I'd change the name to `auto_reconnect` or `reconnect_on_lost`.